glib: Remove ending NUL when converting Variant to OsString.

rodrigorc opened the pull request:
Fixes #620.

sdroege commented:
Makes sense to me but now I'm wondering if maybe we should add `NUL`s when serializing those types into variants. I wonder what C code making use of paths in variants is expecting.

Might make sense bringing that up as a GLib issue if this is not documented anywhere.

rodrigorc commented:
> ...but now I'm wondering if maybe we should add `NUL`s when serializing those types into variants.

I was just thinking about that!

>  I wonder what C code making use of paths in variants is expecting. Might make sense bringing that up as a GLib issue if this is not documented anywhere.

I think it is documented if you look deep enough:
In [GApplication](https://github.com/GNOME/glib/blob/main/gio/gapplication.c#L709):

> * for G_OPTION_ARG_FILENAME, use `^&ay`

With `ay` being an array of bytes, that would be read with [g-variant-get-bytestring](https://www.freedesktop.org/software/gstreamer-sdk/data/docs/latest/glib/glib-GVariant.html#g-variant-get-bytestring), that says:
> If the array does not end with a nul terminator character, the empty string is returned. For this reason, you can always trust that a non-NULL nul-terminated string will be returned by this function.

So my guess is that the NUL byte is mandatory. Should I update this pull-request or create a new one, as this is technically a different issue?

What I don't quite understand here is why the Windows version stores `WCHAR` (`u16`) values instead of bytes. Nothing in the GLib source code that I saw suggests that an array of `WCHAR` should be used there. Maybe I'm missing something?

I don't have a Windows machine around to try, maybe I'll try with Wine and let you know.

jf2048 commented:
That does look like it should probably always get converted to utf8, to match what glib always does. Exposing the utf16 is a rust thing

sdroege commented:
> That does look like it should probably always get converted to utf8, to match what glib always does. Exposing the utf16 is a rust thing

The "local options" in `GApplication` are using a non-UTF-8 string for filename options though. It would be in local encoding (except on Windows where it's always UTF-8?).

> What I don't quite understand here is why the Windows version stores `WCHAR` (`u16`) values instead of bytes. Nothing in the GLib source code that I saw suggests that an array of `WCHAR` should be used there. Maybe I'm missing something?

That was because `OsString` uses `u16` on Windows, but probably does not make much sense considering the above, and should probably always use UTF-8 on Windows. (This means it would have to work similar to `ToGlibPtr` / `FromGlibPtr` for `OsString` / `Path` which also special-cases Windows).

> With `ay` being an array of bytes, that would be read with [g-variant-get-bytestring](https://www.freedesktop.org/software/gstreamer-sdk/data/docs/latest/glib/glib-GVariant.html#g-variant-get-bytestring), that says:

`ay` is a byte array that doesn't have to end on `NUL` but `get_bytestring()` is a helper function that assumes that for "byte strings" (i.e. not byte arrays), AFAIU.

I think your analysis is correct in that regard and it should indeed be assuming / putting a `NUL` at the end for `Path` / `OsString` conversions.

jf2048 commented:
Hmm, yes it does look like the "local encoding" on windows is supposed to be utf8: [g_win32_get_command_line](https://developer-old.gnome.org/glib/stable/glib-Windows-Compatibility-Functions.html#g-win32-get-command-line)

rodrigorc commented:
Ok, I'm trying to fix everything in one go. I've just added a new commit and squashed the PR.

I've added changes to `OsString::from_variant()` and `OsStr::to_variant()`, both for Unix and Windows variants, pushing and popping the NUL byte appropriately.

I'm not totally sure about the lossy conversions in Windows... But then, if you have a Windows file name that is not UTF-8 valid, I think that you have already lost: Rust can handle that with the internal WTF-8 encoding of OsString, but AFAIK glib uses pure UTF-8. The alternative to lossy conversion would be to replace them with an empty string, as Variant does with the missing NUL.

I have tested my example from the original bug report, both in Linux and in Windows (tricky!) and now it seems to work great. (previously Windows failed with something about the type of variant being "aq" instead of "ay").

sdroege commented:
I would go with how it's implemented in `glib/src/translate.rs` in `c_to_path_buf` and `path_to_c`. I mean, you can even call those functions from here after making them `pub(crate)` :)

sdroege commented on glib/src/variant.rs:
```diff
@@ -1230,8 +1234,13 @@ impl FromVariant for std::ffi::OsString {
+        while bytes.last() == Some(&0) {
+            bytes.pop();
+        }
+        let s = String::from_utf8_lossy(&bytes);
```
I.e. here you would panic instead

rodrigorc replied:
Makes sense, I'll use `c_to_path_buf` and `c_to_os_string` instead `PathBuf::FromVariant` calling `OsString::FromVariant`.

sdroege commented on glib/src/variant.rs:
```diff
@@ -1276,9 +1288,11 @@ impl StaticVariantType for std::ffi::OsStr {
-        wide.to_variant()
+        // Just like the Unix variant, except that the internal [u8] of OsSTr is not visible
+        // and the wchar array ([u16]) is not useful, because it should be utf-8 encoded.
+        let mut bytes = self.to_string_lossy().into_owned().into_bytes();
```
And here too

sdroege commented on glib/src/variant.rs:
```diff
@@ -1218,7 +1218,11 @@ impl ToVariant for std::ffi::OsString {
 #[cfg(unix)]
```
The docs above here and also below need to be updated

rodrigorc replied:
But these docs only want about portability. That comment is there even for `any(unix,windows)`, so I guess these now should be just that.

sdroege replied:
They're not really platform-specific anymore but according to the GLib rules. Before there was the `u8` / `u16` difference, now they're just byte arrays in either case and work the same like everything else in GLib on all platforms.

So I think either these portability notes should be removed, or maybe there should be a note of the above (GLib rules, local encoding on UNIX, UTF-8 on Windows).

sdroege commented on glib/src/variant.rs:
```diff
@@ -1218,7 +1218,11 @@ impl ToVariant for std::ffi::OsString {
 #[cfg(unix)]
 impl FromVariant for std::ffi::OsString {
```
And also further above the `Path` conversions need to consider long paths on Windows like in `path_to_c()`

rodrigorc replied:
Calling `path_to_c` literally should just work. I guess. I'll use `os_str_to_c` to convert `OsStr` to avoid that when not requested.

sdroege commented:
Otherwise this looks all good to me!

rodrigorc commented:
I have force-pushed a new version of this PR.

Funnily, using `translate.rs` makes the code shorter: (23 insertions(+), 54 deletions(-)).

sdroege commented on glib/src/variant.rs:
```diff
@@ -1163,7 +1163,10 @@ impl ToVariant for std::path::PathBuf {
 #[cfg(any(unix, windows))]
```
This `cfg`s can go away now. `glib::translate` always provides these helper functions and they're unconditionally used (indirectly) from elsewhere too.

sdroege commented on glib/src/variant.rs:
```diff
@@ -1185,7 +1188,8 @@ impl StaticVariantType for std::path::Path {
     fn to_variant(&self) -> Variant {
-        self.as_os_str().to_variant()
+        let tmp = crate::translate::path_to_c(self);
+        return unsafe { from_glib_none(ffi::g_variant_new_bytestring(tmp.as_ptr() as *const u8)) };
```
```suggestion
        unsafe {
            from_glib_none(ffi::g_variant_new_bytestring(tmp.as_ptr() as *const u8))
        }
```

rodrigorc replied:
Ops! Done.

sdroege commented on glib/src/variant.rs:
```diff
@@ -1250,35 +1243,11 @@ impl StaticVariantType for std::ffi::OsStr {
-        let wide = self.encode_wide().collect::<Vec<u16>>();
-        wide.to_variant()
+        let tmp = crate::translate::os_str_to_c(self);
+        return unsafe { from_glib_none(ffi::g_variant_new_bytestring(tmp.as_ptr() as *const u8)) };
```
```suggestion
        unsafe {
            from_glib_none(ffi::g_variant_new_bytestring(tmp.as_ptr() as *const u8))
        }
```

sdroege commented:
The implementation looks all good to me, just needs a little bit of cleanup and then this can go in.

> Funnily, using translate.rs makes the code shorter: (23 insertions(+), 54 deletions(-)).

That's nice :)

rodrigorc commented:
I've just force-pushed a new version.
Also, now that these functions call into `translate.rs` I've fully separated the impls of `Path/PathBuf` and `OsStr/OsString`.

rodrigorc commented on glib/src/variant.rs:
```diff
@@ -1134,151 +1134,75 @@ impl ToVariant for str {
-        Some(std::ffi::OsString::from_variant(variant)?.into())
+        unsafe {
+            let ptr = ffi::g_variant_get_bytestring(variant.to_glib_none().0);
+            Some(crate::translate::c_to_path_buf(ptr as *const i8))
```
That `c_char` thing is madness!

sdroege replied:
It is indeed. Bad choices made many decades ago that are still haunting us :)
